src/cpp/rtps/transport/shared_mem/RobustInterprocessCondition.hpp

```diff
@@ -0,0 +1,341 @@
+// Copyright 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef _FASTDDS_SHAREDMEM_ROBUST_INTERPROCESS_CONDITION_
+#define _FASTDDS_SHAREDMEM_ROBUST_INTERPROCESS_CONDITION_
+
+#include <boost/interprocess/sync/detail/locks.hpp>
+#include <boost/interprocess/sync/interprocess_mutex.hpp>
+#include <boost/interprocess/sync/interprocess_semaphore.hpp>
+#include <boost/interprocess/sync/scoped_lock.hpp>
+
+namespace eprosima {
+namespace fastdds {
+namespace rtps {
+
+namespace bi = boost::interprocess;
+
+class RobustInterprocessCondition
+{
+public:
+
+    RobustInterprocessCondition()
+        : list_listening_(SemaphoreList::LIST_NULL)
+        , list_free_(MAX_LISTENERS-1)
+    {
+        init_sem_list();
+    }
+
+    ~RobustInterprocessCondition()
+    {
+    }
+
+    /**
+     * If there is a thread waiting on *this, change that
+     * thread's state to ready. Otherwise there is no effect.
+     * @throw boost::interprocess::interprocess_exception on error.
+     */
+    void notify_one()
+    {
+        bi::scoped_lock<bi::interprocess_mutex> lock(semaphore_lists_mutex_);
+
+        auto sem_index = list_listening_.tail();
```

Comment:
I think we should use a FIFO mechanism instead of a LIFO one. Otherwise, a single thread could starve other waiting threads.